### PR TITLE
CHANGE: Removed aliased values in Gamepad enum. (ISXB-543)

### PR DESCRIPTION
**Pull request by adrian-koretski-unity3d**
### Description

Fixing confusing menu when selecting face buttons due to aliased enum values. 

### Changes made

Removed aliased enum values. 

### Testing

Manual testing. 

### Risk

None. 

### Checklist

Before review:

- [ ] Changelog entry added.
    - Explains the change in `Changed`, `Fixed`, `Added` sections.
    - For API change contains an example snippet and/or migration example.
    - JIRA ticket linked, example ([case %<ID>%](https://issuetracker.unity3d.com/product/unity/issues/guid/<ID>)). If it is a private issue, just add the case ID without a link.
    - Jira port for the next release set as "Resolved".
- [ ] Tests added/changed, if applicable.
    - Functional tests `Area_CanDoX`, `Area_CanDoX_EvenIfYIsTheCase`, `Area_WhenIDoX_AndYHappens_ThisIsTheResult`.
    - Performance tests.
    - Integration tests.
- [ ] Docs for new/changed API's.
    - Xmldoc cross references are set correctly.
    - Added explanation how the API works.
    - Usage code examples added.
    - The manual is updated, if needed.

During merge:

- [ ] Commit message for squash-merge is prefixed with one of the list:
    - `NEW: ___`.
    - `FIX: ___`.
    - `DOCS: ___`.
    - `CHANGE: ___`.
    - `RELEASE: 1.1.0-preview.3`.

After merge:

- [ ] Create forward/backward port if needed. If you are blocked from creating a forward port now please add a task to ISX-1444.

### CODE REVIEW

**Comment by unity-cla-assistant** [comment]
[CLA assistant check](https://cla-assistant.cds.internal.unity3d.com/Unity-Technologies/InputSystem?pullRequest=1983) <br/>All committers have signed the CLA.

File: Packages/com.unity.inputsystem/InputSystem/Devices/Gamepad.cs
Comment (ekcoh) [inline Packages/com.unity.inputsystem/InputSystem/Devices/Gamepad.cs]: We cannot remove these from this enum without requiring a major version revision since it would break public API for existing customers (and according to semver), hence this is not a viable way forward.

File: Packages/com.unity.inputsystem/CHANGELOG.md
Comment (ekcoh) [inline Packages/com.unity.inputsystem/CHANGELOG.md]: We do not want to release a major revision due to this change.

**Review by ekcoh — CHANGES_REQUESTED** [review]

**Comment by ekcoh** [comment]
I would recommend considering a fix based off this one but with inlined/combined options for aliased entries? https://github.com/Unity-Technologies/InputSystem/pull/1862

**Comment by ekcoh** [comment]
Seems like @Billreyn  was also considering that unless opinion has shifted, see https://github.com/Unity-Technologies/InputSystem/pull/1862#issuecomment-1971097297

File: Packages/com.unity.inputsystem/InputSystem/Devices/Gamepad.cs
Comment (adrian-koretski-unity3d) [reply Packages/com.unity.inputsystem/InputSystem/Devices/Gamepad.cs]: I hadn't thought of this. I re-added the aliased enum values (although I did not reset them in the other files that used them to move towards north/south/east/west). I also added a slightly modified version of the aliased enum drawer from [#1862](https://github.com/Unity-Technologies/InputSystem/pull/1862) where we can remove values from the drawer.

**Comment by adrian-koretski-unity3d** [comment]
> Seems like @Billreyn was also considering that unless opinion has shifted, see [#1862 (comment)](https://github.com/Unity-Technologies/InputSystem/pull/1862#issuecomment-1971097297)

I believe so. Based on our conversation, we want to display only North/South/East/West

File: Packages/com.unity.inputsystem/CHANGELOG.md
Comment (adrian-koretski-unity3d) [reply Packages/com.unity.inputsystem/CHANGELOG.md]: Should we plan to do so down the line? Perhaps mark the aliased values as depricated (I'm not sure how we go about doing this).

File: Assets/Tests/InputSystem/CoreTests_Actions.cs
Comment (ekcoh) [inline Assets/Tests/InputSystem/CoreTests_Actions.cs]: The fix shouldn't have any impact on existing tests?

Comment (ekcoh) [reply Assets/Tests/InputSystem/CoreTests_Actions.cs]: Hence I wouldn't expect any changes to any tests.

File: Assets/Tests/InputSystem/Plugins/UserTests.cs
Comment (ekcoh) [inline Assets/Tests/InputSystem/Plugins/UserTests.cs]: This shouldn't need to change either.

File: Packages/com.unity.inputsystem/CHANGELOG.md
Comment (ekcoh) [reply Packages/com.unity.inputsystem/CHANGELOG.md]: We basically shouldn't modify the enum at all, we should aim for a completely editor-based "fix" here. There is no functional issue with the current solution, its only confusing users from UX perspective.

**Comment by ekcoh** [comment]
> > Seems like @Billreyn was also considering that unless opinion has shifted, see [#1862 (comment)](https://github.com/Unity-Technologies/InputSystem/pull/1862#issuecomment-1971097297)
> 
> I believe so. Based on our conversation, we want to display only North/South/East/West

Exactly, with emphasis on "display", hence it should be property drawer only changes as I see it.

File: Packages/com.unity.inputsystem/CHANGELOG.md
Comment (ekcoh) [reply Packages/com.unity.inputsystem/CHANGELOG.md]: It makes sense though to file an internal ticket to consider removing aliases for control enums down the line (in a major future release). I have now filed it.

File: Packages/com.unity.inputsystem/InputSystem/Devices/Gamepad.cs
Comment (ekcoh) [inline Packages/com.unity.inputsystem/InputSystem/Devices/Gamepad.cs]: This seems incorrect? Referencing to the same thing

File: Assets/Tests/InputSystem/CoreTests_Events.cs
Comment (ekcoh) [inline Assets/Tests/InputSystem/CoreTests_Events.cs]: Not expecting any changes needed to this file

File: Assets/Tests/InputSystem/CoreTests_State.cs
Comment (ekcoh) [inline Assets/Tests/InputSystem/CoreTests_State.cs]: Not expecting any changes needed to this file

File: Packages/com.unity.inputsystem/InputSystem/Editor/PropertyDrawers/AliasedEnumPropertyDrawer.cs
Comment (ekcoh) [inline Packages/com.unity.inputsystem/InputSystem/Editor/PropertyDrawers/AliasedEnumPropertyDrawer.cs]: Screenshot 2024-09-13 at 11 09 14
Screenshot 2024-09-13 at 11 11 23
I am afraid this is not doing what its supposed to.....

The following things need to be addressed:
- Aliased values should be part of the same entry when the inspector is displayed, e.g. "South / X / A"
- Drop down need to allow all values to be picked, e.g. either "South / X / A" or each one individually would be fine from my perspective.
- At the moment when selecting e.g. South, the value displayed is "North", so something with mapping is incorrect

File: Packages/com.unity.inputsystem/InputSystem/Editor/PropertyDrawers/AliasedEnumPropertyDrawer.cs
Comment (ekcoh) [inline Packages/com.unity.inputsystem/InputSystem/Editor/PropertyDrawers/AliasedEnumPropertyDrawer.cs]: Since the enum is small and we want consistency, maybe skip this base class and just move into specific GamepadButtonPropertyDrawer class and just use a big switch statement to get it right? That would be quite straightforward. Otherwise there needs to be some smartness to get the order of each aliased combination separated by slashes consistent.

**Review by ekcoh — CHANGES_REQUESTED** [review]
Needs a bit more work I believe

File: Assets/Tests/InputSystem/Plugins/UserTests.cs
Comment (adrian-koretski-unity3d) [reply Assets/Tests/InputSystem/Plugins/UserTests.cs]: I believe this is due to my first pass where I removed the aliased values. Will revert

File: Assets/Tests/InputSystem/CoreTests_Actions.cs
Comment (adrian-koretski-unity3d) [reply Assets/Tests/InputSystem/CoreTests_Actions.cs]: I believe this is due to my first pass where I removed the aliased values. Will revert

File: Packages/com.unity.inputsystem/InputSystem/Devices/Gamepad.cs
Comment (adrian-koretski-unity3d) [reply Packages/com.unity.inputsystem/InputSystem/Devices/Gamepad.cs]: I believe this is due to my first pass where I removed the aliased values. Will revert

File: Packages/com.unity.inputsystem/InputSystem/Editor/PropertyDrawers/AliasedEnumPropertyDrawer.cs
Comment (adrian-koretski-unity3d) [reply Packages/com.unity.inputsystem/InputSystem/Editor/PropertyDrawers/AliasedEnumPropertyDrawer.cs]: I'll go over it and make sure it behaves and displays correctly

File: Packages/com.unity.inputsystem/InputSystem/Editor/PropertyDrawers/GpadButtonPropertyDrawer.cs
Comment (lyndon-unity) [inline Packages/com.unity.inputsystem/InputSystem/Editor/PropertyDrawers/GpadButtonPropertyDrawer.cs]: Minor but can you use a full name here rather than abbreviation 
GpadButtonPropertyDrawer ->
GamepadButtonPropertyDrawer

**Review by lyndon-unity — APPROVED** [review]
Code looks reasonable, one minor recommendation, but I haven't tested it in action.
Suggest adding QA or design for a manual verification

File: Packages/com.unity.inputsystem/CHANGELOG.md
Comment (ekcoh) [inline Packages/com.unity.inputsystem/CHANGELOG.md]: This should be reverted as well, suggest to change to something like e.g. "Changed representation of GamepadButton enum values in Inspector to display aliased enum values as a single item to avoid confusion around selection and aliased value display when multiple enum items map to the same numerical value."

File: Packages/com.unity.inputsystem/InputSystem/Editor/PropertyDrawers/GpadButtonPropertyDrawer.cs
Comment (ekcoh) [reply Packages/com.unity.inputsystem/InputSystem/Editor/PropertyDrawers/GpadButtonPropertyDrawer.cs]: Agree

Comment (ekcoh) [reply Packages/com.unity.inputsystem/InputSystem/Editor/PropertyDrawers/GpadButtonPropertyDrawer.cs]: Which also means renaming file and .meta

File: Packages/com.unity.inputsystem/InputSystem/Editor/PropertyDrawers/GpadButtonPropertyDrawer.cs
Comment (ekcoh) [inline Packages/com.unity.inputsystem/InputSystem/Editor/PropertyDrawers/GpadButtonPropertyDrawer.cs]: This assignment could be on line 40 right since overwritten for the special cases?

File: Packages/com.unity.inputsystem/InputSystem/Editor/PropertyDrawers/GpadButtonPropertyDrawer.cs
Comment (ekcoh) [inline Packages/com.unity.inputsystem/InputSystem/Editor/PropertyDrawers/GpadButtonPropertyDrawer.cs]: Could be moved up to the place where it is currently assigned as null?

**Review by ekcoh — APPROVED** [review]
Approving this to not block this PR in any way while I have limited accessability. I didn't run it on this review round so recommend checking functionality in QA pass.

**Comment by Pauliusd01** [comment]
Will check this today or tomorrow morning at the latest

**Review by Pauliusd01 — CHANGES_REQUESTED** [review]
Entering play mode with one of these new values selected breaks inspector and throws errors. Also, the enum value does not match the one in the inspector (See debug log in screenshot, whether that's expected I dunno) 

1726666305 IN-45621_-_Scene_-_Windows,_Mac,_Linux_-_Unity_6_P

File: Packages/com.unity.inputsystem/InputSystem/Editor/PropertyDrawers/GpadButtonPropertyDrawer.cs
Comment (adrian-koretski-unity3d) [reply Packages/com.unity.inputsystem/InputSystem/Editor/PropertyDrawers/GpadButtonPropertyDrawer.cs]: I changed it up a little bit. For the cases where we do not want to enter the enum into the list, I continue the for loop, eliminating the need for null/null check. The switch statement sets an string value which is then entered into the list.

**Comment by adrian-koretski-unity3d** [comment]
> Entering play mode with one of these new values selected breaks inspector and throws errors. Also, the enum value does not match the one in the inspector (See debug log in screenshot, whether that's expected I dunno)
> 
> 1726666305 IN-45621_-Scene-_Windows,_Mac,Linux-_Unity_6_P

The nullref issue should be fixed now. The value being displayed will be one of the aliased enums for that value so this is to be expected.

File: Packages/com.unity.inputsystem/CHANGELOG.md
Comment (Pauliusd01) [inline Packages/com.unity.inputsystem/CHANGELOG.md]: Changelog entry in the wrong version and no bug issue link

**Review by Pauliusd01 — APPROVED** [review]

**Comment by Pauliusd01** [comment]
@adrian-koretski-unity3d I approved too hastily, the link has to be to a public issue tracker not our internal jira lol

**Pull request merged**
